### ISSUE TITLE
Create a mini-MU plugin for the Server-Timing API

### ISSUE THREAD

**Issue by mukeshpanchal27** [10up/wordpress-develop#3]
- Based on the [Performance Lab plugin implementation](https://github.com/WordPress/performance/tree/trunk/server-timing) create MU plugin for Server-Timing API

### Acceptance Criteria

- Add a new folder called `performance` to the `wordpress-develop/tests` directory.
- Create a new folder called `server-timing` and place the necessary files in it.
- Add new step in `performance.yml` action that add MU plugin in docker environment.
```
name: Install MU plugin
run: |
  mkdir ./src/wp-content/mu-plugins
  cp ./tests/performance/server-timing/mu-plugin.php ./src/wp-content/mu-plugins/mu-plugin.php
```

The MU Plugin should record and report the following metrics via the Server-Timing API:

- `wp-before-template`: Time it takes from WordPress core's `timer_start()` call until the end of `template_redirect` action
- `wp-template`: Time it takes from end of `template_redirect` action to start of `shutdown`
- `wp-total`: Sum of both of the above metrics

Note that in order to measure `wp-template` and output it in a header, we need to use an output buffer. That's okay since it'll be running on all tests, thus not skewing data.

[This Gist](https://gist.github.com/felixarntz/de5c697a1a16c2b892634b70216eb6c7) can serve as potential inspiration for the implementation of the MU plugin.

**Comment by joemcgill** [comment]
I'm curious if we could just use the PL plugin outright for now rather than creating a separate MU plugin? We would need to load that plugin via code, but we may not need a totally separate implementation.

**Comment by mukeshpanchal27** [comment]
Thank you for your input, @joemcgill. I agree that we can use the PL plugin instead of the MU plugin for the initial work. @felixarntz Can you share your thoughts on this?

We also use the MU plugin because it is simple to use in GH. We can add one folder to `tests` and load it.

```
- name: Install MU plugin
   run: |
     mkdir ./src/wp-content/mu-plugins
     cp ./tests/performance/mu-plugin.php ./src/wp-content/mu-plugins/mu-plugin.php
```

**Comment by joemcgill** [comment]
I think the mu-plugin idea you have here is fine for an initial approach. I think the important part is that we're able to capture server timing metrics, so whatever tool (e.g., standalone plugin or PL plugin) that helps us do that effectively is fine for now.

**Comment by adamsilverstein** [comment]
> I'm curious if we could just use the PL plugin outright for now rather than creating a separate MU plugin?

👍🏼 we discussed this a bit during design and we decided to build something separate from the PL plugin in part because that way we can keep it very focused.  

> Create a new folder called server-timing and place the necessary files in it.
nit. I think we will want a folder called `plugins` in the `performance` folder and then a `server-timing` plugin folder in there.

**Comment by joemcgill** [comment]
@adamsilverstein do you know if there is already prior art for this mu-plugin that we should use, or will we need to create something from scratch based on the server timing API approach in the PL plugin?

**Comment by adamsilverstein** [comment]
I think we need to build this from scratch based on the pl plugin code.

**Comment by joemcgill** [comment]
It would probably be good to capture the initial server timing metrics we want to capture here as well. The PL plugin is currently capturing these:

* wp-load-alloptions-query
* wp-before-template
* wp-before-template-db-queries

Do we want to measure wp-template and wp-template-db-queries as well? Any others?

**Comment by joemcgill** [comment]
Actually, it looks like @felixarntz [has a gist](https://gist.github.com/felixarntz/de5c697a1a16c2b892634b70216eb6c7) we might be able to use as a starting point here.

**Comment by felixarntz** [comment]
@joemcgill @adamsilverstein I think the most important metrics would be `wp-before-template` and `wp-template`. Potentially we also want to have another metric that is just the sum of these two? That would effectively be the overall WordPress execution time, and IMO it'd be nice to have it exposed as a single value, so that you don't have to always sum up the two partial metrics for that.

The `*-db-queries` ones we could include, can also show interesting results and are generic enough. But not _as_ important as the above.

`wp-load-alloptions-query` is a very specific metric that the PL plugin covers, but that would not really fit into the general set of metrics we're defining for this MVP.

Regarding the code for the Server-Timing "API" in this plugin, I'd suggest keeping it as simple as possible since this won't be an external-facing API like the one in the PL plugin. I'd say something closer to my Gist; maybe that one is even _too_ much of a quick hack, but I think we should keep it simple (e.g. no need for much validation, registration methods etc.).

**Comment by joemcgill** [comment]
Makes sense to me, I'll add the following to the ticket description for now and we can iterate:

* wp-before-template
* wp-template
* wp-total

**Comment by felixarntz** [comment]
@joemcgill I've expanded on the ACs a bit more just to explain what exactly those 3 metrics should measure. I've also added a note on output buffering and the Gist. Please take a look.

From my perspective the ACs are good to proceed.

**Comment by eclarke1** [comment]
PR is approved, so I have moved this over to the 'Approval' column which indicates it is ready to be merged. Once merged, please move to Done.

**Comment by mukeshpanchal27** [comment]
@joemcgill Can i mark this issue closed?

**Comment by joemcgill** [comment]
Resolved by #7.
